jupiterone/resource_user_group_membership.go

```diff
@@ -0,0 +1,306 @@
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/jupiterone/terraform-provider-jupiterone/jupiterone/internal/client"
+)
+
+type UserGroupMembershipResource struct {
+	version string
+	qlient  graphql.Client
+}
+
+// UserGroupMembershipModel is the terraform HCL representation of a user group membership.
+type UserGroupMembershipModel struct {
+	Id      types.String `json:"id,omitempty" tfsdk:"id"`
+	GroupId types.String `json:"groupId,omitempty" tfsdk:"group_id"`
+	Email   types.String `json:"email,omitempty" tfsdk:"email"`
+}
+
+func NewUserGroupMembershipResource() resource.Resource {
+	return &UserGroupMembershipResource{}
+}
+
+// Metadata implements resource.Resource
+func (*UserGroupMembershipResource) Metadata(ctx context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_user_group_membership"
+}
+
+// Schema implements resource.Resource
+func (*UserGroupMembershipResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Description: "A saved JupiterOne User Group Membership.",
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed: true,
+			},
+			"group_id": schema.StringAttribute{
+				Required: true,
+				Description: "The id of the group to add the user to.",
+			},
+			"email": schema.StringAttribute{
+				Required:    true,
+				Description: "The email of the user to add to the group.",
+			},
+		},
+	}
+}
+
+// Configure implements resource.ResourceWithConfigure
+func (r *UserGroupMembershipResource) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	p, ok := req.ProviderData.(*JupiterOneProvider)
+
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected JupiterOneProvider, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+
+		return
+	}
+
+	r.version = p.version
+	r.qlient = p.Qlient
+}
+
+// Create implements resource.Resource
+func (r *UserGroupMembershipResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var data UserGroupMembershipModel
+
+	// Read Terraform plan data into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	_, err := client.InviteUser(
+		ctx,
+		r.qlient,
+		data.Email.ValueString(),
+		data.GroupId.ValueString(),
+	)
+
+	if err != nil {
+		resp.Diagnostics.AddError("failed to create user group membership", err.Error())
+		return
+	}
+
+	tflog.Trace(ctx, "Created user group membership",
+		map[string]interface{}{"groupId": data.GroupId, "email": data.Email})
+
+	data.Id = types.StringValue("placeholder")
```

Comment:
Could you expand on what this is doing/why it is necessary? It's going to set the Id to "placeholder" for all these resources? Is that okay?

Reply:
This is called out in the docs here: https://developer.hashicorp.com/terraform/tutorials/providers-plugin-framework/providers-plugin-framework-acceptance-testing#implement-data-source-id-attribute

I will add a comment for newcomers.

Reply:
I added the comment on the struct itself so that I don't muddy up everywhere I set the value as placeholder.